### ISSUE TITLE
UNION-ing two disparate records, custom prefetched rows?

### ISSUE THREAD

**Issue by mgray88** [groue/GRDB.swift#1005]
### What did you do?

Given three `FetchableRecord`s, the first of which (`RouteRecord`/`RouteStepInfo`) is an index with references to two others, I'm attempting to generate a list of records of two different types (`PickupStepInfo`, `DropOffStepInfo`). I can accomplish this part successfully with a custom SQL statement with the `UNION` operator. The records decode successfully with simple custom `init(row:)`s.

The issue I'm facing is one of the records (`DropOffStepInfo`) requires an `including(all:)` which would require a separate sql query, however it doesn't seem I can finagle that into the `row` initializer the way it would be if the sql query was generated by GRDB's request interface.

Is there someway to accomplish this, or maybe a better way to go about it?

<details>
<summary>RouteRecord table</summary>

```swift
public struct RouteRecord: Codable, FetchableRecord, PersistableRecord, TableRecord {
    public static let databaseTableName: String = "route"

    public let index: Int
    public let cartID: CartID
    public let step: Step

    public init(index: Int, cartID: CartID, step: Step) {
        self.index = index
        self.cartID = cartID
        self.step = step
    }

    public enum Step: Codable {
        case restaurant(RestID)
        case customer

        public init(from decoder: Decoder) throws {
            let container = try decoder.singleValueContainer()
            let value = try container.decode(String.self)
            if value == "cust" {
                self = .customer
            } else {
                self = .restaurant(value)
            }
        }

        public func encode(to encoder: Encoder) throws {
            var container = encoder.singleValueContainer()
            switch self {
            case let .restaurant(restID):
                try container.encode(restID)

            case .customer:
                try container.encode("cust")
            }
        }
    }
}

extension RouteRecord {
    public enum Columns: String, ColumnExpression {
        case index
        case cartID
        case step
    }
}
```
</details>

<details>
<summary>RouteStepInfo</summary>

 ```swift
public enum RouteStepInfo: Hashable {
    case pickup(PickupStepInfo)
    case dropOff(DropOffStepInfo)
}
extension RouteStepInfo: FetchableRecord {
    public init(row: Row) {
        let record = RouteRecord(row: row)
        switch record.step {
        case .restaurant:
            self = .pickup(PickupStepInfo(row: row))
            break

        case .customer:
            self = .dropOff(DropOffStepInfo(row: row))
            break
        }
    }
}
```
</details>

<details>
<summary>PickupStepInfo</summary>

```swift
public struct PickupStepInfo: Decodable, DataStruct, Hashable {
    public let cartID: CartID
    public let restID: RestID
    public let orderNumber: Int
    public let restaurantName: String
    @CodableColor
    public private(set) var color: UIColor
    public let address: String
    public let status: DriverStatus

    public init(
        cartID: CartID,
        restID: RestID,
        orderNumber: Int,
        restaurantName: String,
        color: UIColor,
        address: String,
        status: DriverStatus
    ) {
        self.cartID = cartID
        self.restID = restID
        self.orderNumber = orderNumber
        self.restaurantName = restaurantName
        self.color = color
        self.address = address
        self.status = status
    }

    public enum CodingKeys: String, CodingKey {
        case cartID
        case restID
        case orderNumber
        case restaurantName
        case color
        case address
        case status
    }
}
extension PickupStepInfo: FetchableRecord {
    public init(row: Row) {
        let cartID: CartID = row[RouteRecord.Columns.cartID]
        let restID: RestID = row[RestaurantRecord.Columns.restID]
        let orderNumber: Int = row[RestaurantRecord.Columns.orderNumber]
        let restaurantName: String = row[RestaurantRecord.Columns.restName]
        let color: UIColor = row[ColorMapRecord.Columns.color]
        let address: String = row["address"]
        let status: DriverStatus = row[RestaurantRecord.Columns.status]
        self.init(
            cartID: cartID,
            restID: restID,
            orderNumber: orderNumber,
            restaurantName: restaurantName,
            color: color,
            address: address,
            status: status
        )
    }
}
```
</details>

<details>
<summary>DropOffStepInfo</summary>

```swift
public struct DropOffStepInfo: Decodable, DataStruct, Hashable {
    public let cartID: CartID
    public let restaurants: [Restaurant]
    public let address: String
    public let directions: String?
    public let status: OrderStatus

    public init(
        cartID: CartID,
        restaurants: [Restaurant],
        address: String,
        directions: String?,
        status: OrderStatus
    ) {
        self.cartID = cartID
        self.restaurants = restaurants
        self.address = address
        self.directions = directions
        self.status = status
    }

    public enum CodingKeys: String, CodingKey {
        case cartID
        case restaurants
        case address
        case directions
        case status
    }

    public struct Restaurant: Decodable, DataStruct, Hashable {
        public let orderNumber: Int
        public let restName: String
        @CodableColor
        public private(set) var color: UIColor

        public init(orderNumber: Int, restName: String, color: UIColor) {
            self.orderNumber = orderNumber
            self.restName = restName
            self.color = color
        }

        public enum CodingKeys: String, CodingKey {
            case orderNumber
            case restName
            case color
        }
    }
}
extension DropOffStepInfo: FetchableRecord {
    public init(row: Row) {
        let cartID: CartID = row[RouteRecord.Columns.cartID]
        let restaurants: [Restaurant] = [] // row["restaurants"]
        let address: String = row["address"]
        let directions: String? = row[DeliveryInfoRecord.Columns.directions]
        let status: OrderStatus = row[OrderRecord.Columns.status]
        self.init(
            cartID: cartID,
            restaurants: restaurants,
            address: address,
            directions: directions,
            status: status
        )
    }
}
```
</details>

<details>
<summary>SQL request</summary>

```swift 
try RouteStepInfo.fetchAll(db, sql: """
```
```sql
SELECT
    "route"."index",
    "route"."cartID",
    "route"."step",
    "cart_restaurant"."restID",
    "cart_restaurant"."orderNumber",
    "cart_restaurant"."restName",
    "cart_restaurant"."restStreet" || ' ' || "cart_restaurant"."restZip" AS "address",
    "cart_restaurant"."status",
    "color_map"."color",
    null
FROM "route"
JOIN "cart_restaurant"
ON ("cart_restaurant"."cartID" = "route"."cartID") AND ("cart_restaurant"."restID" = "route"."step")
JOIN "color_map"
ON ("color_map"."cartID" = "cart_restaurant"."cartID") AND ("color_map"."restID" = "cart_restaurant"."restID")
UNION
SELECT
    "route"."index",
    "route"."cartID",
    "route"."step",
    null,
    null,
    null,
    "delivery_info"."street" || ' ' || "delivery_info"."zip" AS "address",
    "order"."status",
    null,
    "delivery_info"."directions"
FROM "route"
JOIN "order"
ON "order"."cartID" = "route"."cartID"
JOIN "delivery_info"
ON "delivery_info"."cartID" = "route"."cartID"
WHERE "route"."step" = 'cust'
ORDER BY "route"."index"
```
```swift
""")
```
</details>

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** 5.8.0
**Installation method:** CocoaPods
**Xcode version:** 12.5
**Swift version:** 5.4
**Platform(s) running GRDB:** iOS
**macOS version running Xcode:** 11.4

**Comment by groue** [comment]
Hello @mgray88,

Please tell me if I understand your issue well:

1. You are trying to fetch an array of `DropOffStepInfo`, where each `DropOffStepInfo` has itself an array of related elements: `restaurants`.

2. In the context of GRDB associations, you'd like to be able to get your `[DropOffStepInfo]` by calling a `fetchAll` method on a some base request extended with an `including(all:)` step. Something like:

    ```swift
    let infos = someBaseRequest
        .including(all: restaurants)
        .asRequest(of: DropOffStepInfo.self)
        .fetchAll(db)
    ```

3. Unfortunately, due to the lacking support for `UNION` in the Swift sql builder, you have to write your base request with raw SQL 🤓 

4. And raw SQL request do not support `including(all:)` 😨 

5. You don't know how to feed each `DropOffStepInfo` with its restaurants! 😭 

---

In the response below, I assume the above description is true.

First, this description is accurate: there indeed is no current support for `UNION` in the query builder, and there is no way to extend a raw SQL request with `including(all:)`. Those are both possible improvements. But since they are not there today, and those are not trivial tasks, let's solve your problem in your application.

As you correctly write, each `including(all:)` performs one extra SQL request, and then gathers the results of both requests. You can perform the exact same thing in your application code, using the same algorithm used by GRDB. I'll assume that the restaurants are associated through `cartID` (you'll have to adapt the sample code this is incorrect).

```swift
// Same as DropOffStepInfo, but without the restaurants
struct PartialDropOffStepInfo: FetchableRecord {
    var cartID: CartID
    var address: String
    var directions: String?
    var status: OrderStatus
    
    init(row: Row) {
        let cartID: CartID = row[RouteRecord.Columns.cartID]
        let address: String = row["address"]
        let directions: String? = row[DeliveryInfoRecord.Columns.directions]
        let status: OrderStatus = row[OrderRecord.Columns.status]
        self.init(
            cartID: cartID,
            address: address,
            directions: directions,
            status: status
        )
    }
}

// Fetch partial step infos
let partialInfos = try /* fetch PartialDropOffStepInfo from SQL */

// Fetch their restaurants
let cartIDs = partialInfos.map(\.cartID)
let restaurants = try Restaurant
    .filter(cartIDs.contains(Column("cartID")))
    .fetchAll(db)

// Group restaurant by cartID
let restaurantsByCartID = Dictionary(
    grouping: restaurants,
    by: \.cartID)

// Build the array of complete DropOffStepInfo
let dropOffStepInfos = partialInfos.map { partialInfo -> DropOffStepInfo in
    let restaurants = restaurantsByCartID[partialInfo.cartID, default: []]
    return DropOffStepInfo(partialInfo, restaurants)
}
```

**Comment by mgray88** [comment]
Hi @groue, thanks for the quick response!

> You are trying to fetch an array of DropOffStepInfo

I'm actually trying to fetch an array of `RouteStepInfo` ordered by `RouteRecord.index` where `RouteStepInfo` is either `.pickup(PickupStepInfo)` or `.dropOff(DropOffStepInfo)`. The given sql query actually feeds `RouteStepInfo(row:)` where it's then further decoded to individual types.

> where each DropOffStepInfo has itself an array of related elements: restaurants.

But this part is true, which is where I was struggling. I chose UNION to reduce the amount of requests I would have to make otherwise.

It looks like though with a little tweaking your solution may be the only way to accomplish this

**Comment by groue** [comment]
It looks like you found your solution. Please reopen if you have any other questions about the topic of this issue.

**Comment by mgray88** [comment]
Hi @groue, yep that was immensely helpful, and I was able get everything working. Almost. Up until now, I haven't used every value returned from the request. I'm now attempting to access the value at `"delivery_info"."directions"` from the second half of the `UNION` statement. Only... the name of the column returned in the `Row` is `"null"`. Since the value expected for that column is a nullable string, it hasn't caused any issues till this point.

I've tried forcing a column alias in the query (`AS "directions"`). I've also tried moving that part of the select statement elsewhere, (in between two other columns, all the way at the beginning). I've tried adding an extra `null` selected column at the end. All to no avail.

The name of that column, and only that column is always `"null"`. Any insights?
